tools/scalajsenv.js

```diff
@@ -864,14 +881,21 @@ initArray(
   const componentBase = componentData.arrayBase || componentData;
   const arrayDepth = componentData.arrayDepth + 1;
 
+  const typeTag = (((1 << 8) | arrayDepth) << 23) | componentBase.typeTag;
+
+  ArrayClass.prototype.$typeTag = typeTag;
+  ScalaJS.ClassData[typeTag] = this;
```

Comment:
We could store array class tags in a different array, that array would still need to be sparse though (because of nested arrays). But it would not hurt the performance of all non-array objects.